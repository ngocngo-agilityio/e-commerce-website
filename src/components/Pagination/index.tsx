'use client';

// Libs
import { usePathname, useRouter, useSearchParams } from 'next/navigation';
import { memo, useCallback } from 'react';
import { Center } from '@chakra-ui/react';
import { ArrowLeftIcon, ArrowRightIcon } from '@chakra-ui/icons';
import {
  PaginationPrevious,
  Pagination as PaginationA,
  PaginationPageGroup,
  PaginationPage,
  PaginationNext,
  usePagination,
} from '@ajna/pagination';

// Utils
import { getSearchParams, updateSearchParams } from '@utils';

interface Props {
  pagesQuantity: number;
}

const Pagination = ({ pagesQuantity }: Props): JSX.Element => {
  const pathname = usePathname();
  const searchParams = useSearchParams();
  const { replace } = useRouter();

  const { page } = getSearchParams(searchParams);
  const selectedPage = parseInt(page) || 1;

  const { currentPage, setCurrentPage, pagesCount, pages } = usePagination({
    pagesCount: pagesQuantity,
    initialState: { currentPage: selectedPage },
  });

  const handlePageChange = useCallback(
    (page: number) => {
      setCurrentPage(page);

      const updatedParams = updateSearchParams(
        searchParams,
        'page',
        page.toString(),
      );

      replace(`${pathname}?${updatedParams.toString()}`, { scroll: false });
    },
    [pathname, replace, searchParams, setCurrentPage],
  );

  return (
    <Center>
      <PaginationA
        pagesCount={pagesCount}
        currentPage={currentPage}
        onPageChange={handlePageChange}
      >
        <>
          <PaginationPrevious
            variant="outline"
            bg="pagination.color"
            color="pagination.bg"
            _hover={{ bg: 'pagination.color', color: 'pagination.hoverBg' }}
            _disabled={{
              bg: 'pagination.color',
              color: 'pagination.bg',
            }}
          >
            <ArrowLeftIcon />
          </PaginationPrevious>

          <PaginationPageGroup style={{ display: 'flex', gap: '10px' }}>
            <>
              {pages.map((page) => (
                <PaginationPage
                  key={`Pagination_page_${page}`}
                  page={page}
                  w="25px"
                  h="25px"
                  bg="pagination.bg"
                  color="pagination.color"
                  fontSize="md"
                  fontWeight="normal"
                  _hover={{ bg: 'pagination.hoverBg' }}
                  _current={{
                    w: '25px',
                    h: '25px',
                    bg: 'pagination.activeBg',
                    color: 'pagination.color',
                    fontSize: 'md',
                    fontWeight: 'normal',

                    _hover: {
                      bg: 'pagination.hoverBg',
                    },
                  }}
                />
              ))}
            </>
          </PaginationPageGroup>

          <PaginationNext
            variant="outline"
            bg="pagination.color"
            color="pagination.bg"
            _hover={{ bg: 'pagination.color', color: 'pagination.hoverBg' }}
            _disabled={{ bg: 'pagination.color', color: 'pagination.bg' }}
          >
            <ArrowRightIcon />
          </PaginationNext>
        </>
      </PaginationA>
    </Center>
  );
};

export default memo(Pagination);
