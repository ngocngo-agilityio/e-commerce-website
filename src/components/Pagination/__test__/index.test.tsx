import { render, fireEvent } from '@testing-library/react';
import Pagination from '..';

const mockReplace = jest.fn();
jest.mock('next/navigation', () => ({
  usePathname: jest.fn().mockReturnValue('/example/path'),
  useRouter: jest.fn().mockReturnValue({ replace: () => mockReplace }),
  useSearchParams: jest.fn().mockReturnValue(new URLSearchParams('page=1')),
}));

describe('Pagination component', () => {
  it('renders pagination with correct number of pages', () => {
    const { container } = render(<Pagination pagesQuantity={5} />);

    expect(container).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('calls setCurrentPage and replace on page change', () => {
    const { getByText } = render(<Pagination pagesQuantity={5} />);

    fireEvent.click(getByText('2'));
  });
});
