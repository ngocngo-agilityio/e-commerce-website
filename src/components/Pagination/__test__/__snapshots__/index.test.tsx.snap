// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component renders pagination with correct number of pages 1`] = `
<div>
  <div
    class="css-gmuwbf"
  >
    <button
      aria-disabled="true"
      aria-label="Previous page"
      class="chakra-button pagination-previous css-6ggxn3"
      disabled=""
      type="button"
    >
      <svg
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <g
          fill="currentColor"
        >
          <path
            d="M10.416,12a2.643,2.643,0,0,1,.775-1.875L20.732.584a1.768,1.768,0,0,1,2.5,2.5l-8.739,8.739a.25.25,0,0,0,0,.354l8.739,8.739a1.768,1.768,0,0,1-2.5,2.5l-9.541-9.541A2.643,2.643,0,0,1,10.416,12Z"
          />
          <path
            d="M.25,12a2.643,2.643,0,0,1,.775-1.875L10.566.584a1.768,1.768,0,0,1,2.5,2.5L4.327,11.823a.25.25,0,0,0,0,.354l8.739,8.739a1.768,1.768,0,0,1-2.5,2.5L1.025,13.875A2.643,2.643,0,0,1,.25,12Z"
          />
        </g>
      </svg>
    </button>
    <ol
      class="chakra-stack pagination-page-group css-1x343b"
      style="display: flex; gap: 10px;"
    >
      <li
        class="css-k008qs"
      >
        <button
          aria-current="true"
          aria-disabled="false"
          aria-label="Current page, page 1"
          class="chakra-button pagination-page css-19soc5h"
          type="button"
        >
          1
        </button>
      </li>
      <li
        class="css-k008qs"
      >
        <button
          aria-current="false"
          aria-disabled="false"
          aria-label="Go to page 2"
          class="chakra-button pagination-page css-1dl2lk3"
          type="button"
        >
          2
        </button>
      </li>
      <li
        class="css-k008qs"
      >
        <button
          aria-current="false"
          aria-disabled="false"
          aria-label="Go to page 3"
          class="chakra-button pagination-page css-1dl2lk3"
          type="button"
        >
          3
        </button>
      </li>
      <li
        class="css-k008qs"
      >
        <button
          aria-current="false"
          aria-disabled="false"
          aria-label="Go to page 4"
          class="chakra-button pagination-page css-1dl2lk3"
          type="button"
        >
          4
        </button>
      </li>
      <li
        class="css-k008qs"
      >
        <button
          aria-current="false"
          aria-disabled="false"
          aria-label="Go to page 5"
          class="chakra-button pagination-page css-1dl2lk3"
          type="button"
        >
          5
        </button>
      </li>
    </ol>
    <button
      aria-disabled="false"
      aria-label="Next page"
      class="chakra-button pagination-next css-6ggxn3"
      type="button"
    >
      <svg
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <g
          fill="currentColor"
        >
          <path
            d="M13.584,12a2.643,2.643,0,0,1-.775,1.875L3.268,23.416a1.768,1.768,0,0,1-2.5-2.5l8.739-8.739a.25.25,0,0,0,0-.354L.768,3.084a1.768,1.768,0,0,1,2.5-2.5l9.541,9.541A2.643,2.643,0,0,1,13.584,12Z"
          />
          <path
            d="M23.75,12a2.643,2.643,0,0,1-.775,1.875l-9.541,9.541a1.768,1.768,0,0,1-2.5-2.5l8.739-8.739a.25.25,0,0,0,0-.354L10.934,3.084a1.768,1.768,0,0,1,2.5-2.5l9.541,9.541A2.643,2.643,0,0,1,23.75,12Z"
          />
        </g>
      </svg>
    </button>
  </div>
</div>
`;
